streaming/streaming.go

```diff
@@ -274,6 +292,98 @@ func getSourceStreamConfig(srcCfg *datastreampb.SourceConfig, sourceProfile prof
+	pubsubClient, err := pubsub.NewClient(ctx, projectID)
+	if err != nil {
+		return fmt.Errorf("pubsub client can not be created: %v", err)
+	}
+	defer pubsubClient.Close()
+	fmt.Println("Created pubsub client...")
+
+	topicObj, err := pubsubClient.CreateTopic(ctx, streamingCfg.PubsubCfg.TopicId)
+	if err != nil {
+		return fmt.Errorf("pubsub topic could not be created: %v", err)
+	}
+
+	_, err = pubsubClient.CreateSubscription(ctx, streamingCfg.PubsubCfg.SubscriptionId, pubsub.SubscriptionConfig{
+		Topic:             topicObj,
+		AckDeadline:       time.Minute * 10,
+		RetentionDuration: time.Hour * 24 * 7,
+	})
+	if err != nil {
+		return fmt.Errorf("pubsub subscription could not be created: %v", err)
+	}
+	fmt.Printf("Successfully created pubsub topic and subscription")
+
+	//Creating datastream client to fetch the gcs bucket using target profile.
+	dsClient, err := datastream.NewClient(ctx)
+	if err != nil {
+		return fmt.Errorf("datastream client can not be created: %v", err)
+	}
+	defer dsClient.Close()
+
+	// Fetch the GCS path from the destination connection profile.
+	dstProf := fmt.Sprintf("projects/%s/locations/%s/connectionProfiles/%s", projectID, datastreamCfg.DestinationConnectionConfig.Location, datastreamCfg.DestinationConnectionConfig.Name)
+	res, err := dsClient.GetConnectionProfile(ctx, &datastreampb.GetConnectionProfileRequest{Name: dstProf})
+	if err != nil {
+		return fmt.Errorf("could not get connection profiles: %v", err)
+	}
+	gcsProfile := res.Profile.(*datastreampb.ConnectionProfile_GcsProfile).GcsProfile
+	bucketName := gcsProfile.Bucket
+	prefix := gcsProfile.RootPath + datastreamCfg.DestinationConnectionConfig.Prefix
+	if prefix != "" && prefix[len(prefix)-1] != '/' {
+		prefix = prefix + "/"
+	}
+	prefix = prefix + "data/"
+	if prefix[0] == '/' {
+		prefix = prefix[1:]
+	}
+
+	storageClient, err := storage.NewClient(ctx)
+	if err != nil {
+		return fmt.Errorf("GCS storage client can not be created: %v", err)
+	}
+	defer storageClient.Close()
+
+	notification := storage.Notification{
+		TopicID:          streamingCfg.PubsubCfg.TopicId,
+		TopicProjectID:   projectID,
+		PayloadFormat:    storage.JSONPayload,
+		ObjectNamePrefix: prefix,
+	}
+
+	createdNotification, err := storageClient.Bucket(bucketName).AddNotification(ctx, &notification)
+	if err != nil {
+		return fmt.Errorf("GCS Notification could not be created: %v", err)
+	}
+	fmt.Printf("Successfully created pubsub topic, subscription and notification for bucket %s.\n", bucketName)
+	streamingCfg.PubsubCfg.NotificationId = createdNotification.ID
+	streamingCfg.PubsubCfg.BucketName = bucketName
+	storeGeneratedPubsubResources(conv, streamingCfg.PubsubCfg, projectID, streamingCfg.DataShardId)
+	return nil
+}
+
+func storeGeneratedPubsubResources(conv *internal.Conv, pubsubCfg internal.PubsubCfg, project string, dataShardId string) {
```

Comment:
Storing of resources happens in LaunchDataflow function. It is able to store all the Datastream resources because the IDs for all the datastream resources can be specified by us while creating them. So, we are able to pre-generate Datastream resource IDs into streamingCfg. And LaunchDataflow function can get those IDs from streamingCfg and store them. But Notification config which we create on the GCS bucket for this pubsub fix, doesn't allow us to provide custom ID. we need to store the ID which that create call returns. Instead of returning the notification Id from here and then pass it till it reaches LaunchDataflow method, I passed the conv object to CreatePubsubResources method and stored the Ids.

Reply:
Understood, thanks for explaining.
I think in this case we should still refactor the generated resource storing logic to either of the following - 

Approach 1:

1. Store the resource information in a local object after each resource creation.
2. Return the local object back to the calling function where `conv` is present and store each resource in it as it generates.

Approach 2:

1. Move the current storage logic from `LaunchDataflow` to somewhere else so that we can still more all the generated values together.

Reply:
Done.